Source/WebKit/UIProcess/haiku/BackingStoreHaiku.cpp

```diff
@@ -29,34 +33,103 @@
 #if USE(HAIKU)
 
 #include "UpdateInfo.h"
-#include <WebCore/NotImplemented.h>
+
+#include "WebCore/IntRect.h"
+#include "WebCore/ShareableBitmap.h"
+#include <Rect.h>
 
 namespace WebKit {
+using namespace WebCore;
 
-static const Seconds s_scrollHysteresisDuration { 300_ms };
+// BackingStore stores and updates a bitmap of the rendered webpage.
 
 BackingStore::BackingStore(const WebCore::IntSize& size, float deviceScaleFactor)
     : m_size(size)
     , m_deviceScaleFactor(deviceScaleFactor)
-    , m_scrolledHysteresis([this](PAL::HysteresisState state) { if (state == PAL::HysteresisState::Stopped) m_scrollSurface = nullptr; }, s_scrollHysteresisDuration)
+    , m_bitmap(BRect(0, 0, size.width() * deviceScaleFactor, size.height() * deviceScaleFactor), B_RGBA32, true)
+    , m_view(m_bitmap.Bounds(), "BackingStore", 0, 0)
 {
+	m_bitmap.AddChild(&m_view);
 }
 
-BackingStore::~BackingStore() = default;
+BackingStore::~BackingStore()
+{
+    m_bitmap.RemoveChild(&m_view);
+}
 
-void BackingStore::paint(PlatformPaintContextPtr cr, const WebCore::IntRect& rect)
+void BackingStore::paint(BView* into, const WebCore::IntRect& rect)
 {
-    notImplemented();
+    // Paint the contents of our bitmap into the BView.
+    into->PushState();
+    into->SetDrawingMode(B_OP_COPY);
+    into->DrawBitmap(&m_bitmap, rect, rect);
+    into->PopState();
+    // TODO: Would SetViewBitmap work instead? We probably would only need
+    // to call it once from the WebView. Is it faster?
```

Comment:
There is also the option to use DrawBitmapAsync, which sends the draw command to app_server but doesn't wait for it to complete. It has similar implications (you shouldn't start drawing to the bitmap until drawing it is comlete)